feat: Absolute stage position controls

gselzer opened the pull request:
This PR is designed to resolve #332.

gselzer commented:
I am especially interested in the reviewers' opinions on a036481, which fixes/improves the testing of setting absolute position. Allowing error does not make sense to me, but it is also necessary for the tests to pass (although the positional error is much smaller than the bounds I wrote)...

tlambert03 commented:
Thanks @gselzer, the code looks good to me.  My main concern, and the reason I backed down from adding this in https://github.com/pymmcore-plus/pymmcore-widgets/pull/334 is that this introduces a very real danger to damage hardware.  programmatic control of motors, particularly when the objective is focused on a sample, is one of the main causes of expensive damage.  I did like @marktsuchida's suggestion in https://github.com/pymmcore-plus/pymmcore-widgets/pull/334#issuecomment-2233371248 that this functionality might be hidden behind an additional button and dialog that would prevent anyone doing this by accident.  I also think it would be great if that dialog also had a big "HALT" button that would stop and existing stage movement.  (the diSPIM plugin has one of those in micro-manager and i love it and use it all the time)

gselzer commented:
> My main concern, and the reason I backed down from adding this in https://github.com/pymmcore-plus/pymmcore-widgets/pull/334 is that this introduces a very real danger to damage hardware. programmatic control of motors, particularly when the objective is focused on a sample, is one of the main causes of expensive damage.

This is a great point - happy to refactor this PR so that we can resolve your concerns.

> I did like @marktsuchida's suggestion in [#334 (comment)](https://github.com/pymmcore-plus/pymmcore-widgets/pull/334#issuecomment-2233371248) that this functionality might be hidden behind an additional button and dialog that would prevent anyone doing this by accident.

This is one option. A second, simpler option is just an opt-in toggle to enable these spinboxes - otherwise they could just perform like the label that you had before. I actually thought about adding that as a part of this PR, but opted for simplicity. Happy to delegate to your and @marktsuchida's intuition here.

> I also think it would be great if that dialog also had a big "HALT" button that would stop and existing stage movement.

There's already [code for that](https://github.com/pymmcore-plus/pymmcore-widgets/blob/7668eda444e8dfff5a6ef665165b541901fb6cd3/src/pymmcore_widgets/control/_stage_widget.py#L72) although we could maybe create a smaller button with an icon instead of text, and put it alongside these spinboxes.

tlambert03 commented:
> A second, simpler option is just an opt-in toggle to enable these spinboxes - otherwise they could just perform like the label that you had before.

I'm good with that option 👍... I wonder about spacing though, the X/Y case is already pretty wide horizontally, and has 4 checkboxes below. 

<img width="226" alt="Screenshot 2024-11-08 at 3 39 23 PM" src="https://github.com/user-attachments/assets/358ce8ba-8042-4e9a-a336-048c6e232636">

just an annoying aesthetic issue :) if you can come up with a nice look for it, let's go with that option

tlambert03 commented:
hey @gselzer, just checking in here. let me know if you want any more ideas or just need time to implement.  I'd like to see a screenshot of how the buttons look when added

gselzer commented:
> hey @gselzer, just checking in here. let me know if you want any more ideas or just need time to implement. I'd like to see a screenshot of how the buttons look when added

Oh! Sorry, lost track of this - I'll take a look tomorrow once I have some time

gselzer commented:
@tlambert03 with 5034319 the UI looks like this:

![image](https://github.com/user-attachments/assets/11d384fa-a095-4fa3-9eee-c12819e837a8)

Thoughts?

tlambert03 commented:
I was expecting a button somewhere that toggles whether the XY positions are editable in the first place (with the default being not editable, the user should have to do some explicit action to enter a mode where the stage will move in response to number entry).

marktsuchida commented:
> * We'd then have to add another control - I feel a bit bad about all of the clutter I've already added 😅

That was one of the motivations for proposing that an additional dialog open. Alternatively, the text can be uneditable by default, but a small popup could open on right-clicking and selecting "Edit...". Perhaps.

fdrgsp commented:
>  the text can be uneditable by default, but a small popup could open on right-clicking and selecting "Edit...

I like this idea!

marktsuchida commented:
Of course this small popup would need to stay open (until something else is clicked?) if we want to allow people to edit it multiple times without re-clicking. Also, perhaps something like "Enter Coordinates..." would be a better name than "Edit...".

gselzer commented:
@fdrgsp I do like that implementation - are you able to push that to this branch?
Couple points:
* Should we tether the two spin boxes together on a 2-axis stage, such that both become enabled/disabled together?
* I like @marktsuchida's suggestion of  `Edit...` for the menu text, but then it should probably change to something else when enabled. Maybe `Enable Editing`/`Disable Editing`?

gselzer commented:
@fdrgsp thank you so much for contributing, in cf9942a I cleaned up your changes a bit, and added some tests. Notably, I:
* Tethered the spin boxes so enabling one enables both.
* Made the menu action toggle, so no need to change text.
* More minor changes.
Lemme know if you see any issues

fdrgsp commented:
@gselzer I like this solution but after having a quick chat with @tlambert03 I do realize that it might be a little hard for a user to know that there is the option to enable manual editing of the coordinates. 

What about having a menu on top of the widget? It might be easier to "discover"...


https://github.com/user-attachments/assets/1c4678cc-0149-4fc6-86d0-c9306de96d94

gselzer commented:
> @gselzer I like this solution but after having a quick chat with @tlambert03 I do realize that it might be a little hard for a user to know that there is the option to enable manual editing of the coordinates.

You know, this reminds me of the discussion that we had in #336 - specifically "once you notice they're editable, you know it forever."

I have a *very weak* preference for the current behavior as:
* It requires one fewer control.
* The menu is physically closer to the widgets it affects (but this is likely easily resolved).
* This behavior maybe *shouldn't* be discovered and used unless you're a poweruser who really knows what they're doing.
All of these reasons are pretty minor, so I'm content with being overruled here if you guys like it better this way.
